### PR TITLE
nixos/cloudflared: init

### PR DESCRIPTION
**Pull request by bbigras**
###### Description of changes

Restarting work from #123432 (which is clearly a made up number)

Can be used like this:

```nix
{
  services.cloudflared = {
    enable = true;
    tunnels = {
      "00000000-0000-0000-0000-000000000000" = {
        credentialsFile = "${config.sops.secrets.cloudflared-creds.path}";
        ingress = {
          "*.domain1.com" = {
            service = "http://localhost:80";
            path = "/*.(jpg|png|css|js)";
          };
          "*.domain2.com" = "http://localhost:80";
        };
        default = "http_status:404";
      };
    };
  };
}
```

TODO
- [ ]  `extraConfig` (tunnel and ingress)
- [ ] `extraArgs` (`cloudflared tunnel run --args`)
- [ ] nixos test (offline test ingress rules from the config with `cloudflared tunnel ingress rule`)

maybe (or in another PR)
- [ ] validate config using `cloudflared tunnel ingress validate` while building (like nginx does)
- [ ] maybe prevent restarting cloudflared while running nixos-rebuild to prevent closing connections. I have no idea if cloudflared supports config reloading but I think they recommend starting a replicas before stopping the service while updating the config).

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bbigras** [comment]
I added options. Tell me if I missed some. I ~~stole~~ copied some descriptions from the documentation.

**Comment by curbengh** [comment]
cloudflared is also a DoH client. The [official doc](https://developers.cloudflare.com/1.1.1.1/encryption/dns-over-https/dns-over-https-client/) doesn't mention any option. I found following example in the [pi-hole doc](https://docs.pi-hole.net/guides/dns/cloudflared/#automatic-way):

``` yml
proxy-dns: true
proxy-dns-port: 5053
proxy-dns-upstream:
  - https://1.1.1.1/dns-query
  - https://1.0.0.1/dns-query
```

**Comment by baduhai** [comment]
@bbigras is still being worked on? I'd love to see it in time for 22.11. If you're no longer working on it, I'll take a shot at it.

**Comment by j2ghz** [comment]
@bbigras Can this be added as-is? The TODO seems to be mostly about extra features, not anything blocking.

**Comment by bbigras** [comment]
Yeah, let's add the extra stuff later.

Please review.

**Comment by boozedog** [comment]
Hi @bbigras is this still progressing? Anything I can help with?

**Comment by bbigras** [comment]
> Anything I can help with?

You can review. We'll try to get it merged as-is. and improve it later.

**Comment by bbigras** [comment]
Oh yeah, I think I need to fix the DocBook thing.

**Comment by bbigras** [comment]
Thanks!!

File: nixos/modules/services/networking/cloudflared.nix [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+              service = "http://localhost:80";
+            };
+          };
+          default = "http_status:404";
+        };
+      };
+    };
+  };
+
+  config = {
```
Comment (winterqt) [inline nixos/modules/services/networking/cloudflared.nix:261]: This isn't guarded by `cfg.enable` ([h/t](https://matrix.to/#/!kjdutkOsheZdjqYmqp%3Anixos.org/%24hlIEO5iBVw7bgtAmc9xtvHNLb8lzfbzbET7bhYAIjEE)).

Comment (winterqt) [reply nixos/modules/services/networking/cloudflared.nix:261]: Fixed in #206422.

**Pull request merged**